### PR TITLE
Upgrade sccache to v0.8.2 for CPU targets

### LINKED ISSUE #121559: [aarch64 linux ci] missing sccache binaries for aarch64 linux platform
### 🐛 Describe the bug

Currently pytorch CI build infrastructure maintains sccache binaries only for x86 architecture. We need the aarch64 linux binaries to speed up the aarch64 ci workflows.  Currently aarch64 linux workflow is crashing if sccache is enabled because x86 binaries are pulled.

```
sccache_epilogue' EXIT
++ [[ -n 1 ]]
++ echo 'Skipping sccache server initialization, setting environment variables'
Skipping sccache server initialization, setting environment variables
++ export SCCACHE_IDLE_TIMEOUT=0
++ SCCACHE_IDLE_TIMEOUT=0
++ export SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log
++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log
++ export RUST_LOG=sccache::server=error
++ RUST_LOG=sccache::server=error
++ sccache --zero-stats
.ci/pytorch/common-build.sh: line 47: /opt/cache/bin/sccache: cannot execute binary file: Exec format error
```

### Versions

aarch64 linux runner:
```
ami-id: ami-0b1c2e7eafac77408
instance-id: i-0fbadc66663ad3bdc
instance-type: t4g.2xlarge
system info Linux ip-10-0-42-211.ec2.internal 4.14.336-[25](https://github.com/pytorch/pytorch/actions/runs/8164152112/job/22318870156#step:4:27)6.559.amzn2.aarch64 #1 SMP Tue Feb 13 09:50:21 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux
```

cc @malfet

### PR DESCRIPTION
**Pull request by malfet**
This essentially reverts https://github.com/pytorch/pytorch/pull/95997 but switches to builds from source to official mozilla's sccache repo for CPU builds, except PCH one, see https://github.com/pytorch/pytorch/issues/139188
- Define `SCCACHE_REGION` for the jobs that needs it. 
- Enable aarch64 builds to use sccache, which allows one to do incremental rebuilds under 10 min, see https://github.com/pytorch/pytorch/actions/runs/11565944328/job/32197278296

Fixes https://github.com/pytorch/pytorch/issues/121559

### CODE REVIEW

**Review by atalman — APPROVED** [review]
LGTM

File: .github/workflows/_linux-build.yml [hunk 1]
```diff
@@ -155,15 +156,21 @@ jobs:
           OUR_GITHUB_JOB_ID: ${{ steps.get-job-id.outputs.job-id }}
           HUGGING_FACE_HUB_TOKEN: ${{ secrets.HUGGING_FACE_HUB_TOKEN }}
         run: |
+          # Fetch aws credential from IMDs
+          eval $(python3 -c 'import boto3; import json; c=boto3.Session().get_credentials().get_frozen_credentials()._asdict(); print(f"export AWS_ACCESS_KEY_ID=\"%s\"" % c["access_key"]); print(f"export AWS_SECRET_ACCESS_KEY=\"%s\"" % c["secret_key"]); print(f"export AWS_SESSION_TOKEN=\"%s\"" % c["token"])')
```
Comment (seemethere) [inline .github/workflows/_linux-build.yml:160]: I feel like if were at the point where we're writing something like this we should probably just break it out into its own separate script

Comment (malfet) [reply .github/workflows/_linux-build.yml:160]: Putting it in a script will force everyone to rebase when this is merged. But let me create a script ahead of time

**Comment by huydhn** [comment]
IIRC, the upstream branch didn't seem to fully work with nvcc like our ancient fork when I tried this a year ago.  I attempted to submit a PR to Mozilla with the additional changes but my Rust knowledge was too limit to make it work.  May be upstream is good now, let's see

**Comment by malfet** [comment]
> May be upstream is good now, let's see

I needed to make a tweak to our wrapper of wrappers and it seems to be working, but indeed, let's see...

File: .ci/docker/common/install_cache.sh [hunk 2]
```diff
@@ -36,18 +36,25 @@ if [ -n "$ROCM_VERSION" ]; then
...
+  esac
 fi
 chmod a+x /opt/cache/bin/sccache
 
 function write_sccache_stub() {
   # Unset LD_PRELOAD for ps because of asan + ps issues
   # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589
-  printf "#!/bin/sh\nif [ \$(env -u LD_PRELOAD ps -p \$PPID -o comm=) != sccache ]; then\n  exec sccache $(which $1) \"\$@\"\nelse\n  exec $(which $1) \"\$@\"\nfi" > "/opt/cache/bin/$1"
+  if [ $1 == "gcc"]; then
+    printf "#!/bin/sh\nif [ \"$1\" = \"-E\" ]; then\n  exec $(which $1)  \"\$@\"\nelif [ \$(env -u LD_PRELOAD ps -p \$PPID -o comm=) != sccache ]; then\n  exec sccache $(which $1) \"\$@\"\nelse\n  exec $(which $1) \"\$@\"\nfi" > "/opt/cache/bin/$1"
```
Comment (seemethere) [inline .ci/docker/common/install_cache.sh:54]: Would a heredoc be better instead?

https://linuxize.com/post/bash-heredoc/#basic-heredoc-examples

I feel like it's a bit unreadable here what this exactly does

**Comment by huydhn** [comment]
@malfet I see you are re-opening this one, so I want to share an anecdote that @jeanschmidt and I have discovered when we are testing AWS A100 w.r.t sccache.

On A100 perf job, fbgemm_gpu is built during the testing phase.  I assume that building fbgemm_gpu requires PT to be there, like vision and audio.  Here is an example https://github.com/pytorch/pytorch/actions/runs/11175526034/job/31068763034#step:21:1652.  The problem is that the build takes 1 hour to finish (on a buffy GCP A100 runner no less)

```
2024-10-04T08:26:26.1083862Z Building wheels for collected packages: fbgemm-gpu
2024-10-04T09:25:40.8352695Z   Building wheel for fbgemm-gpu (setup.py) ... done
2024-10-04T09:25:41.5534095Z Created wheel for fbgemm-gpu: filename=fbgemm_gpu-0.4.1rc0.post421-cp310-cp310-linux_x86_64.whl size=252664091 sha256=c32955755e985c9aef7242f43e73dc1f764cf867eba489d36f5caa90820c7be4
```

On AWS A100, sccache just timed out trying to compile the package.  I disable sccache there https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/common_utils.sh#L195-L200 and the build takes half an hour less (weird)

I'm not sure if this is an issue on how fbgemm_gpu is built or if our ancient version of sccache doesn't work here.  

Also there are some wip on official sccache to support nvcc https://github.com/mozilla/sccache/pull/2247 (not yet merged though)

**Comment by trxcllnt** [comment]
@huydhn public sccache releases already support caching nvcc compilations, and have since (at least) v0.3.3. My https://github.com/mozilla/sccache/pull/2247 PR adds support for [distributing](https://github.com/mozilla/sccache/blob/main/docs/DistributedQuickstart.md) `nvcc` compilations to a build cluster. Presently, sccache only supports running `nvcc` compilations locally.

**Comment by huydhn** [comment]
> @huydhn public sccache releases already support caching nvcc compilations, and have since (at least) v0.3.3. My [mozilla/sccache#2247](https://github.com/mozilla/sccache/pull/2247) PR adds support for [distributing](https://github.com/mozilla/sccache/blob/main/docs/DistributedQuickstart.md) `nvcc` compilations to a build cluster. Presently, sccache only supports running `nvcc` compilations locally.

I see.  Thank you very much for the clarification.  For more context, we are maintaining an ancient fork of sccache at https://github.com/pytorch/sccache with some custom nvcc support.  We definitely want to ditch that the use https://github.com/mozilla/sccache.  There are problem here and there with the move blocking us from doing.

**Comment by malfet** [comment]
I'll restrict using newer sccache to just CPU build as somehow it starts to crash when newer sccache is used

File: .github/actions/build-android/action.yml [hunk 3]
```diff
@@ -42,11 +42,15 @@ runs:
         PR_NUMBER: ${{ github.event.pull_request.number }}
         SHA1: ${{ github.event.pull_request.head.sha || github.sha }}
         SCCACHE_BUCKET: ossci-compiler-cache-circleci-v2
+        SCCACHE_REGION: us-east-1
+        AWS_DEFAULT_REGION: us-east-1
```
Comment (malfet) [inline .github/actions/build-android/action.yml:46]:
```suggestion
```

**Comment by malfet** [comment]
@pytorch merge -f "Let's test in trunk!!!"

**Pull request closed without merging**